lib/matrix/matrix.rb

```diff
@@ -179,17 +152,85 @@ def each(which = :all) # :yield: e
     #     #    3 at 1, 0
     #     #    4 at 1, 1
     #
-    def each_with_index
-      raise NotSupportedError unless block_given?
+    def each_with_index(which = :all) # :yield: e, row, column
+      return to_enum :each_with_index, which unless block_given?
+      case which
+      when :all
+        each_all{|i, j| yield self[i, j], i, j}
+      when :diagonal
+        each_diagonal{|i, j| yield self[i, j], i, j}
+      when :off_diagonal
+        each_off_diagonal{|i, j| yield self[i, j], i, j}
+      when :lower
+        each_lower{|i, j| yield self[i, j], i, j}
+      when :strict_lower
+        each_strict_lower{|i, j| yield self[i, j], i, j}
+      when :strict_upper
+        each_strict_upper{|i, j| yield self[i, j], i, j}
+      when :upper
+        each_upper{|i, j| yield self[i, j], i, j}
+      else
+        raise ArgumentError, "expected #{which.inspect} to be one of :all, :diagonal, :off_diagonal, :lower, :strict_lower, :strict_upper or :upper"
+      end
+      self
+    end
 
+    def each_all(&block)
```

Comment:
This methods don't return items of the matrix -> they aren't `each`.
I think, that they can be extracted into a special module (may be named `Utils` or `TraversalSequence`), if `row_count` and `column_count` put in parameters.